### PR TITLE
Story2

### PR DESCRIPTION
**Pull request by HiredMussel**

### CODE REVIEW

File: src/Entities/StageEntity.php [hunk 1]
```diff
@@ -92,4 +92,14 @@ public function getStageDeleted() : int
...
         return $this->deleted;
     }
+
+    public function setOptions(array $options)
+    {
+        $this->options = $options;
+    }
+
+    public function getOptions():?array
+    {
```
Comment (ashleycoles) [inline src/Entities/StageEntity.php:102]: Dockblocks!

File: src/Entities/StageEntity.php [hunk 2]
```diff
@@ -92,4 +92,14 @@ public function getStageDeleted() : int
     {
         return $this->deleted;
     }
+
+    public function setOptions(array $options)
+    {
+        $this->options = $options;
+    }
+
+    public function getOptions():?array
```
Comment (ashleycoles) [inline src/Entities/StageEntity.php:101]: space after the :

File: src/Entities/StageEntity.php [hunk 3]
```diff
@@ -92,4 +92,14 @@ public function getStageDeleted() : int
     {
         return $this->deleted;
     }
+
+    public function setOptions(array $options)
```
Comment (ashleycoles) [inline src/Entities/StageEntity.php:96]: DOnt think this ever gets used

File: src/Models/StageModel.php [hunk 4]
```diff
@@ -133,6 +151,69 @@ public function updateAllStages(array $stages) : bool
...
+     */
+    public function deleteAllOptions(int $stageId) : bool
+    {
+        $query = $this->db->prepare("UPDATE `options` SET `deleted` = '1' WHERE `stageId` = :stageId");
+        $query->bindParam(':stageId', $stageId);
+        return $query->execute();
+    }
+
+
     public function getDB() : \PDO
```
Comment (ashleycoles) [inline src/Models/StageModel.php:217]: Not your code but please delete this utter monstrosity of a method.

File: src/ViewHelpers/StagesPageViewHelper.php [hunk 5]
```diff
@@ -18,18 +21,60 @@ public static function displayStages($stages)
...
+                    $result .= '<form data-id="" class="optionAddForm ">';
+                }
+                $result .= '<input type="text" class="optionAddTitle col-xs-10" ';
+                $result .= 'placeholder="Type the name of your new option"/>';
+                $result .= '<input type="submit" class="optionAddSubmit btn-success" data-stageid="';
+                $result .= $stage->getStageId().'" value="Submit">';
+                $result .= '</form>';
+                $result .= '</div>';
+                $result .= '</div>';
+                $result .= '</div>';
```
Comment (ashleycoles) [inline src/ViewHelpers/StagesPageViewHelper.php:66]: I *think* you have an extra closing div

File: public/js/stagesPage.js [hunk 6]
```diff
@@ -4,6 +4,99 @@ const editResponse = document.getElementById('editResponse');
...
 const createNewResponse = document.getElementById('createNewResponse');
 const deleteButtons = document.querySelectorAll('.delete');
+const optionButtons = document.querySelectorAll('.toggleEditOptions');
+const optionEditSubmits = document.querySelectorAll('.optionEditSubmit');
+const optionEdits = document.querySelectorAll('.optionEdit');
+const optionDeletes = document.querySelectorAll('.optionDelete');
+const optionAddSubmits = document.querySelectorAll('.optionAddSubmit');
+const optionAddTitle = document.querySelectorAll('.optionAddTitle');
+const optionEditTitle = document.querySelectorAll('.optionEditTitle');
+const optionTitle = document.querySelectorAll('.optionTitle');
```
Comment (ashleycoles) [inline public/js/stagesPage.js:14]: Never used

File: public/js/stagesPage.js [hunk 7]
```diff
@@ -4,6 +4,99 @@ const editResponse = document.getElementById('editResponse');
...
+    optionEditSubmit.addEventListener('click', async (e) => {
+        e.preventDefault()
+        let optionTitle = e.target.parentElement.querySelector('.optionEditTitle').value
+        let optionId = e.target.dataset.optionid
+        let data = {
+            'optionTitle': optionTitle,
+            'optionId': optionId
+        }
+        await sendRequest('api/editStageOption', 'PUT', data)
+        window.location.reload(true)
```
Comment (ashleycoles) [inline public/js/stagesPage.js:37]: passing in true is deprecated, should work without it

File: public/js/stagesPage.js [hunk 8]
```diff
@@ -4,6 +4,99 @@ const editResponse = document.getElementById('editResponse');
...
+const optionDeletes = document.querySelectorAll('.optionDelete');
+const optionAddSubmits = document.querySelectorAll('.optionAddSubmit');
+const optionAddTitle = document.querySelectorAll('.optionAddTitle');
+const optionEditTitle = document.querySelectorAll('.optionEditTitle');
+const optionTitle = document.querySelectorAll('.optionTitle');
+const optionsContainers = document.querySelectorAll('.optionsContainer');
+const optionEditForms = document.querySelectorAll('.optionTableForm');
+
+// Set up the modal for deleting the final option
+const modal=document.querySelector('.modalContainer');
```
Comment (ashleycoles) [inline public/js/stagesPage.js:19]: spaces around =

File: public/js/stagesPage.js [hunk 9]
```diff
@@ -4,6 +4,99 @@ const editResponse = document.getElementById('editResponse');
...
+const closeModal = document.querySelector('.closeModal');
+closeModal.addEventListener('click', (e) => {
+    e.preventDefault();
+    window.location.reload(true);
```
Comment (ashleycoles) [inline public/js/stagesPage.js:23]: true is deprecated

Comment (ashleycoles) [reply public/js/stagesPage.js:23]: (when passing into reload)

File: public/js/stagesPage.js [hunk 10]
```diff
@@ -4,6 +4,99 @@ const editResponse = document.getElementById('editResponse');
...
+                    method: 'DELETE',
+                    headers: {
+                        'Content-Type': 'application/json'
+                    },
+                    body: JSON.stringify(data)
+                });
+                window.location.reload(true);
+            })
+        } else {
+            window.location.reload(true);
```
Comment (ashleycoles) [inline public/js/stagesPage.js:76]: also remove the trues here

File: public/js/stagesPage.js [hunk 11]
```diff
@@ -4,6 +4,99 @@ const editResponse = document.getElementById('editResponse');
...
+            "data": []
+        };
+        inputs.forEach((input) => {
+            result.data.push({
+                'title': input.value,
+                'stageId': stageId
+            })
+        })
+        await sendRequest('/api/addStageOption', 'POST', result)
+        window.location.reload(true)
```
Comment (ashleycoles) [inline public/js/stagesPage.js:97]: and here

File: public/js/stagesPage.js [hunk 12]
```diff
@@ -4,6 +4,99 @@ const editResponse = document.getElementById('editResponse');
...
+        let result = {
+            "data": []
+        };
+        inputs.forEach((input) => {
+            result.data.push({
+                'title': input.value,
+                'stageId': stageId
+            })
+        })
+        await sendRequest('/api/addStageOption', 'POST', result)
```
Comment (ashleycoles) [inline public/js/stagesPage.js:96]: Why is this using the sendRequest function whereas your others are using fetch?

File: src/Controllers/AddStageOptionController.php [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+                        $data = [
+                            'success' => true,
+                            'message' => 'Option added successfully',
+                            'data' => []
+                        ];
+                        $statusCode = 200;
+                    } else {
+                        $data['message'] = 'Error adding to database';
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (HiredMussel) [reply src/Controllers/AddStageOptionController.php:64]: Yes

File: src/Controllers/DeleteStageController.php [hunk 14]
```diff
@@ -45,25 +45,28 @@ public function __invoke(Request $request, Response $response, array $args)
...
                     $stageData = $this->stageModel->getStageById($requestData['id']);
                     if ($stageData) {
-                        if ($this->stageModel->deleteStage($requestData['id'])) {
-                            $data = [
-                                'success' => true,
-                                'message' => 'Stage has been deleted successfully.',
-                                'data' => []
-                            ];
-                            $statusCode = 200;
+                        if (empty($stageData->getOptions())) {
```
[the remark this answers is not in the document]
Comment (HiredMussel) [reply src/Controllers/DeleteStageController.php:51]: Bug was with getStageById not properly populating the options field, resulting in the controller's logic evaluating the contents of the options field as empty. This has been resolved.

File: src/Controllers/DeleteAllStageOptionsController.php [hunk 15]
```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Portal\Controllers;
+
+use Portal\Abstracts\Controller;
+use Portal\Models\StageModel;
+use Psr\Http\Message\ServerRequestInterface as Request;
+use Psr\Http\Message\ResponseInterface as Response;
+use Portal\Entities\OptionsEntity;
```
[the remark this answers is not in the document]
Comment (HiredMussel) [reply src/Controllers/DeleteAllStageOptionsController.php:9]: Apparently not. VS Code hadn't greyed it out, but removing it doesn't seem to have broken anything.

File: src/Entities/StageEntity.php (hunk 3 above)
Comment (HiredMussel) [reply src/Entities/StageEntity.php:96]: Yes, it is

File: src/Models/StageModel.php [hunk 16]
```diff
@@ -133,8 +156,65 @@ public function updateAllStages(array $stages) : bool
...
+     * @return bool
+     */
+    public function deleteOption(int $optionId) : bool
+    {
+        $query = $this->db->prepare("UPDATE `options` SET `deleted` = '1' WHERE `id` = :optionId");
+        $query->bindParam(':optionId', $optionId);
+        return $query->execute();
+    }
+
+    public function getOptionsByStageId(int $stageId) : array
```
Comment (ashleycoles) [inline src/Models/StageModel.php:206]: Docblock!

File: public/js/stagesPage.js [hunk 17]
```diff
@@ -4,6 +4,99 @@ const editResponse = document.getElementById('editResponse');
 const newStageForm = document.getElementById('newStageForm');
 const createNewResponse = document.getElementById('createNewResponse');
 const deleteButtons = document.querySelectorAll('.delete');
+const optionButtons = document.querySelectorAll('.toggleEditOptions');
+const optionEditSubmits = document.querySelectorAll('.optionEditSubmit');
+const optionEdits = document.querySelectorAll('.optionEdit');
+const optionDeletes = document.querySelectorAll('.optionDelete');
+const optionAddSubmits = document.querySelectorAll('.optionAddSubmit');
+const optionAddTitle = document.querySelectorAll('.optionAddTitle');
+const optionEditTitle = document.querySelectorAll('.optionEditTitle');
```
Comment (ashleycoles) [inline public/js/stagesPage.js:13]: optionAddTitle and optionEditTitle also appear to be unused

**Pull request merged**